### PR TITLE
Add timeseries charts in homepage

### PR DESCRIPTION
**Pull request by frozenhelium**

### CODE REVIEW

File: src/hooks/useSizeTracking.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        };
+    });
+
+    useEffect(() => {
+        const resizeObserver = new ResizeObserver((entries) => {
+            const entry = entries.at(0);
+            const contentRect = entry?.contentRect;
+
+            if (contentRect) {
+                setSize({
```
Comment (tnagorra) [inline src/hooks/useSizeTracking.ts:20]: Let's throttle this to 60fps

File: src/hooks/useSizeTracking.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { useEffect, useState } from 'react';
+
+function useSizeTracking(ref: React.RefObject<HTMLElement | SVGSVGElement>, use = true) {
```
Comment (tnagorra) [inline src/hooks/useSizeTracking.ts]: Wouldn't disable=false be better?

File: src/utils/chart.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    x: number;
+    y: number;
+}
+
+export interface Boundary {
+    min: number;
+    max: number;
+}
+
+export function getScaleFunction(
```
Comment (tnagorra) [inline src/utils/chart.ts:11]: Let's add tests

File: src/utils/chart.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        },
+    };
+}
+
+export function getDatesSeparatedByYear(startDate: Date, endDate: Date) {
+    const currentDate = new Date(startDate);
+    currentDate.setDate(1);
+
+    const targetDate = new Date(endDate);
+    targetDate.setDate(1);
```
Comment (tnagorra) [inline src/utils/chart.ts]: It's better to also reset the hours, minutes and seconds to zero

File: src/utils/chart.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    const dates = [];
+
+    while (currentDate.getTime() < targetDate.getTime()) {
+        dates.push(new Date(currentDate));
+        currentDate.setFullYear(currentDate.getFullYear() + 1);
```
Comment (tnagorra) [inline src/utils/chart.ts:63]: It's better to also set Hours, Minutes and Seconds

File: src/utils/chart.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    targetDate.setDate(1);
+
+    const dates = [];
+
+    while (currentDate.getTime() <= targetDate.getTime()) {
+        dates.push(new Date(currentDate));
+        currentDate.setMonth(currentDate.getMonth() + 1);
+    }
+
+    return dates;
```
Comment (tnagorra) [inline src/utils/chart.ts]: It's better to also set Hours, Minutes and Seconds

Comment (tnagorra) [reply src/utils/chart.ts]: We should instead use the functions we are using in other projects

File: src/views/Home/ActiveOperationMap/index.tsx [hunk 7]
```diff
@@ -312,26 +311,29 @@ function ActiveOperationMap(props: Props) {
...
+                                        valueType="number"
                                     />
                                     <TextOutput
                                         value={+appeal.amount_requested}
                                         description={strings.operationPopoverAmountRequested}
+                                        valueType="number"
                                     />
                                     <TextOutput
-                                        value={appeal.amount_funded}
+                                        value={Math.round(+appeal.amount_funded)}
```
Comment (tnagorra) [inline src/views/Home/ActiveOperationMap/index.tsx]: Any reason we are rounding this?

File: src/views/Home/Charts/index.tsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    if (isNotDefined(year)) {
+        return (
+            <YearlyChart
+                onYearClick={setYear}
+            />
+        );
+    }
+
+    return (
+        <MontlyChart
```
Comment (tnagorra) [inline src/views/Home/Charts/index.tsx]:
```suggestion
        <MonthlyChart
```

File: src/views/Home/Charts/YearlyChart/index.tsx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function YearlyChart(props: Props) {
+    const { onYearClick } = props;
+    const strings = useTranslation(i18n);
+
+    const dateList = useMemo(
+        () => getDatesSeparatedByYear(startDate, endDate),
+        [],
+    );
```
Comment (tnagorra) [inline src/views/Home/Charts/YearlyChart/index.tsx]: Shouldn't we move this outside the component

File: src/views/Home/Charts/YearlyChart/index.tsx [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+    const [activePointKey, setActivePointKey] = useState<string>(
+        getFormattedKey(dateList[dateList.length - 1]),
```
Comment (tnagorra) [inline src/views/Home/Charts/YearlyChart/index.tsx]: Let's use function to set state here

File: src/views/Home/Charts/YearlyChart/index.tsx [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            {pending && <BlockLoading className={styles.loading} />}
+            {!pending && (
+                <>
+                    <TimelineChart
+                        className={styles.timelineChart}
+                        timePoints={dateList}
+                        dataKeys={['dref', 'emergencyAppeal']}
+                        valueSelector={(dataKey, date) => (
+                            combinedData?.[dataKey]?.[getFormattedKey(date)]?.count ?? 0
+                        )}
```
Comment (tnagorra) [inline src/views/Home/Charts/YearlyChart/index.tsx]: Let's not inline functions on component jsx

File: src/views/Home/Charts/TimelineChart/index.tsx [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    );
+
+    const handleRectMouseClick: React.MouseEventHandler<SVGRectElement> = useCallback((e) => {
+        const key = (e.target as HTMLElement).getAttribute('data-key');
+        if (key) {
+            onTimePointClick(key);
+        }
+    }, [onTimePointClick]);
+
+    const handleRectMouseEnter: React.MouseEventHandler<SVGRectElement> = useCallback((e) => {
```
Comment (tnagorra) [inline src/views/Home/Charts/TimelineChart/index.tsx:176]: We should define the type the argument `e`

File: src/views/Home/Charts/TimelineChart/index.tsx [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+            return [
+                chartData,
+                yAxisData,
+            ] as const;
+        },
+        [bounds, dataKeys, timePoints, valueSelector],
+    );
+
+    const handleRectMouseClick: React.MouseEventHandler<SVGRectElement> = useCallback((e) => {
```
Comment (tnagorra) [inline src/views/Home/Charts/TimelineChart/index.tsx:169]: We should define the type the argument `e`

File: src/views/Home/Charts/TimelineChart/index.tsx [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                    key: getFormattedKey(timePoint.date),
+                    x,
+                    yValues: listToMap(
+                        timePoint.dataValues,
+                        (datum) => datum.key,
+                        (datum) => yScale(datum.value),
+                    ),
+                    rect: {
+                        start: rectStart,
+                        width: Math.max(0, rectEnd - rectStart),
```
Comment (tnagorra) [inline src/views/Home/Charts/TimelineChart/index.tsx]: Let's use bound from fujs

File: src/views/Home/Charts/MonthlyChart/index.tsx [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                        valueSelector={(dataKey, date) => (
+                            combinedData?.[dataKey]?.[getFormattedKey(date)]?.count ?? 0
+                        )}
+                        classNameSelector={(dataKey) => dataKeyToClassNameMap[dataKey]}
+                        activePointKey={activePointKey}
+                        onTimePointClick={setActivePointKey}
+                        xAxisFormatter={(date) => date.toLocaleString(
+                            undefined,
+                            { month: 'short' },
+                        )}
```
Comment (tnagorra) [inline src/views/Home/Charts/MonthlyChart/index.tsx]: Let's not inline functions on render

File: src/views/Home/Charts/MonthlyChart/index.tsx [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+            heading={heading}
+            withHeaderBorder
+        >
+            {pending && <BlockLoading className={styles.loading} />}
+            {!pending && (
+                <>
+                    <TimelineChart
+                        className={styles.timelineChart}
+                        timePoints={dateList}
+                        dataKeys={['dref', 'emergencyAppeal']}
```
Comment (tnagorra) [inline src/views/Home/Charts/MonthlyChart/index.tsx]: We need to discuss about this

File: src/views/Home/Charts/MonthlyChart/index.tsx [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+    );
+
+    const [activePointKey, setActivePointKey] = useState<string>(
+        getFormattedKey(dateList[0]),
+    );
+
+    const queryParams = {
+        model_type: 'appeal',
+        start_date: encodeDate(new Date(year, 0, 1)),
+        end_date: encodeDate(new Date(year, 11, 31)),
```
Comment (tnagorra) [inline src/views/Home/Charts/MonthlyChart/index.tsx:86]: Maybe we should write helpers for this? It's always confusing to use `0` for first month

File: src/views/Home/Charts/MonthlyChart/index.tsx [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+const dateFormatter = new Intl.DateTimeFormat(
+    undefined,
+    {
+        month: 'long',
+    },
+);
+
+const getFormattedKey = (dateFromProps: string | Date) => {
+    const date = new Date(dateFromProps);
```
Comment (tnagorra) [inline src/views/Home/Charts/MonthlyChart/index.tsx:46]: Let's try not to use `new Date` on string. Let's discuss about this.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/hooks/useSizeTracking.ts (hunk 1 above)
Comment (frozenhelium) [reply src/hooks/useSizeTracking.ts:20]: Added a TODO comment

File: src/utils/chart.ts (hunk 3 above)
Comment (frozenhelium) [reply src/utils/chart.ts:11]: Added test for some, added TODO comment for others

File: src/views/Home/Charts/TimelineChart/index.tsx (hunk 12 above)
Comment (frozenhelium) [reply src/views/Home/Charts/TimelineChart/index.tsx:176]: Already defined in type of callback

File: src/views/Home/Charts/TimelineChart/index.tsx (hunk 13 above)
Comment (frozenhelium) [reply src/views/Home/Charts/TimelineChart/index.tsx:169]: Already defined in type of callback

File: src/views/Search/EmergencyTable/index.tsx [hunk 19]
```diff
@@ -99,7 +99,7 @@ function EmergencyTable(props: Props) {
             'funding_requirements',
             strings.searchEmergencyTableFundingRequirements,
             (emergency) => Number(emergency.funding_requirements),
-            { suffix: ' CHF' },
+            { unit: ' CHF' },
```
Comment (tnagorra) [inline src/views/Search/EmergencyTable/index.tsx]: We need to see if we need to leave a space infront of CHF.

File: src/components/NumberOutput/index.tsx [hunk 20]
```diff
@@ -1,188 +1,95 @@
...
-        }
+            options.useGrouping = !separatorHidden;
+            options.maximumFractionDigits = 2;
 
-        // Convert number to add separator
-        if (isDefined(separator) && fraction === 0) {
-            const withSeparator = addSeparator(num, separator);
-            if (withSeparator) {
-                formattedNumber = withSeparator;
+            if (Math.abs(value) > 1000) {
```
Comment (tnagorra) [inline src/components/NumberOutput/index.tsx]:
```suggestion
            if (Math.abs(value) >= 1000) {
```

File: src/views/Home/Charts/TimelineChart/index.tsx [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+            }
+
+            const timePointsWithData = timePoints.map((timePoint) => {
+                const dataValues = dataKeys.map((dataKey) => ({
+                    key: dataKey,
+                    value: valueSelector(dataKey, timePoint),
+                }));
+
+                const valueList = dataValues.map((value) => value.value);
+                const maxValue = valueList.length === 0 ? 0 : Math.max(...valueList);
```
Comment (tnagorra) [inline src/views/Home/Charts/TimelineChart/index.tsx:82]: We might get "-Infinity" as value, do we handle that?

File: src/views/Home/Charts/TimelineChart/index.tsx [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+                return {
+                    date: timePoint,
+                    numMonths: timePoint.getFullYear() * 12 + timePoint.getMonth(),
+                    dataValues,
+                    maxValue,
+                };
+            });
+
+            const maxValueList = timePointsWithData.map((datum) => datum.maxValue);
+            const maxValue = maxValueList.length === 0 ? 0 : Math.max(...maxValueList);
```
Comment (tnagorra) [inline src/views/Home/Charts/TimelineChart/index.tsx:93]: We might get "-Infinity" as value, do we handle that?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/views/Home/Charts/TimelineChart/index.tsx (hunk 21 above)
Comment (frozenhelium) [reply src/views/Home/Charts/TimelineChart/index.tsx:82]: lol

**Pull request merged**